mimalloc-sys/src/lib.rs

```diff
@@ -9,21 +15,124 @@ extern "C" {
     pub fn mi_calloc(count: size_t, size: size_t) -> *mut c_void;
     pub fn mi_realloc(p: *mut c_void, newsize: size_t) -> *mut c_void;
     pub fn mi_expand(p: *mut c_void, newsize: size_t) -> *mut c_void;
-    pub fn mi_posix_memalign(ptr: *mut *mut c_void, alignment: size_t, size: size_t) -> c_int;
+    pub fn mi_posix_memalign(
+        ptr: *mut *mut c_void,
+        alignment: size_t,
+        size: size_t,
+    ) -> c_int;
     pub fn mi_aligned_alloc(alignment: size_t, size: size_t) -> *mut c_void;
     pub fn mi_free(p: *mut c_void);
     pub fn mi_malloc_size(p: *const c_void) -> size_t;
     pub fn mi_malloc_usable_size(p: *const c_void) -> size_t;
+    pub fn mi_memalign(alignment: size_t, size: size_t) -> *mut c_void;
+
+    pub fn mi_strdup(s: *const c_char) -> *mut c_char;
+    pub fn mi_strndup(s: *const c_char, n: size_t) -> *mut c_char;
+    pub fn mi_realpath(fname: *const c_char, resolved_name: *mut c_char) -> *mut c_char;
+
+    // mimalloc implementations of non-standard or uncommon functions from various oses.
+
+    pub fn mi_cfree(p: *mut c_void);
+    pub fn mi__expand(p: *mut c_void, newsize: size_t) -> *mut c_void;
+    pub fn mi_valloc(size: size_t) -> *mut c_void;
+    pub fn mi_pvalloc(size: size_t) -> *mut c_void;
+
+    pub fn mi_reallocarray(
+        p: *mut c_void,
+        count: size_t,
+        size: size_t,
+    ) -> *mut c_void;
+
+    pub fn mi_aligned_recalloc(
+        p: *mut c_void,
+        newcount: size_t,
+        size: size_t,
+        alignment: size_t,
+    ) -> *mut c_void;
+
+    pub fn mi_aligned_offset_recalloc(
+        p: *mut c_void,
+        newcount: size_t,
+        size: size_t,
+        alignment: size_t,
+        offset: size_t,
+    ) -> *mut c_void;
+
+    pub fn mi_wcsdup(s: *const c_ushort) -> *mut c_ushort;
+    pub fn mi_mbsdup(s: *const c_uchar) -> *mut c_uchar;
+
+    pub fn mi_dupenv_s(
+        buf: *mut *mut c_char,
+        size: *mut size_t,
+        name: *const c_char,
+    ) -> c_int;
+    pub fn mi_wdupenv_s(
+        buf: *mut *mut c_ushort,
+        size: *mut size_t,
+        name: *const c_ushort,
+    ) -> c_int;
+
+    pub fn mi_free_size(p: *mut c_void, size: size_t);
+    pub fn mi_free_size_aligned(
+        p: *mut c_void,
+        size: size_t,
+        alignment: size_t,
+    );
+    pub fn mi_free_aligned(p: *mut c_void, alignment: size_t);
+
+    // FIXME: Don't expose these, These should be `extern "C unwind"` or
```

Comment:
I'm not sure what we should do here. These are explicitly non-noexcept (and noexcept versions exist).

It's worth noting that while most of mimalloc's functions are declared in the source as noexcept, it's missing from a few (the heap api, which is actually the whole reason I added support for the rest D:).

I did a little digging (it's wonderful how small and readable mimalloc is) and they definitely don't throw, so I assume it's a just mistake, but it's possible that that indicates there's some other way they can unwind.

Reply:
Submit a PR upstream that makes these `noexcept`, and once that's merged, we can add these here.